### PR TITLE
[spanmetricsprocessor] Validate duplicate dimensions at start

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

**Description:** <Describe what has changed.>

This pr adds a check at the start of span metrics processor for possible label duplication of the reserved dimensions and user-configured dimensions.

**Link to tracking Issue:** This fixes the problem mentioned in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2772#issuecomment-805347326

**Testing:** Added a unit test for the new `validateDimensions` method.

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: processor/spanmetricsprocessor/processor.go [hunk 1]
```diff
@@ -379,3 +416,33 @@ func (p *processorImp) cache(serviceName string, span pdata.Span, k metricKey) {
 		p.metricKeyToDimensions[k] = buildDimensionKVs(serviceName, span, p.dimensions)
 	}
 }
+
+// copied from prometheus-go-metric-exporter
+// sanitize replaces non-alphanumeric characters with underscores in s.
+func sanitize(s string) string {
```
Comment (albertteoh) [inline processor/spanmetricsprocessor/processor.go:422]: Probably worth copying the [tests from prometheusexporter](https://github.com/open-telemetry/opentelemetry-collector/blob/3e138d73b3c44928045046211da421810703fd15/exporter/prometheusexporter/sanitize_test.go) too.

File: processor/spanmetricsprocessor/factory.go [hunk 2]
```diff
@@ -47,5 +47,5 @@ func createDefaultConfig() configmodels.Processor {
 }
 
 func createTraceProcessor(_ context.Context, params component.ProcessorCreateParams, cfg configmodels.Processor, nextConsumer consumer.Traces) (component.TracesProcessor, error) {
-	return newProcessor(params.Logger, cfg, nextConsumer), nil
```
Comment (albertteoh) [inline processor/spanmetricsprocessor/factory.go:50]: Thanks!

File: processor/spanmetricsprocessor/processor_test.go [hunk 3]
```diff
@@ -493,3 +496,76 @@ func TestBuildKey(t *testing.T) {
...
+			name: "we skip the case if the dimension name is the same after sanitization",
+			dimensions: []Dimension{
+				{Name: "http_status_code"},
+			},
+			hasError: false,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			err := validateDimensions(tc.dimensions)
+			if tc.hasError {
```
Comment (albertteoh) [inline processor/spanmetricsprocessor/processor_test.go:564]: We could remove `hasError` and replace with `if tc.expectedErr != nil` (or "" if we decide to pass the err string instead).

File: processor/spanmetricsprocessor/processor_test.go [hunk 4]
```diff
@@ -493,3 +496,76 @@ func TestBuildKey(t *testing.T) {
...
+				assert.Error(t, err, tc.expectedErr)
```
Comment (albertteoh) [inline processor/spanmetricsprocessor/processor_test.go:565]: Please consider using `assert.EqualError` and replace `expectedErr error` with `expectedErr string`. I suspect this assertion is passing, but on a different error to the `expectedErr`.

File: processor/spanmetricsprocessor/processor.go [hunk 5]
```diff
@@ -98,6 +102,10 @@ func newProcessor(logger *zap.Logger, config configmodels.Exporter, nextConsumer
 		}
 	}
 
+	if err := validateDimensions(pConfig.Dimensions); err != nil {
+		return nil, err
```
Comment (albertteoh) [inline processor/spanmetricsprocessor/processor.go:106]: Please add a test to cover this line.

File: processor/spanmetricsprocessor/processor.go [hunk 6]
```diff
@@ -121,6 +129,35 @@ func mapDurationsToMillis(vs []time.Duration, f func(duration time.Duration) flo
...
+			return fmt.Errorf("duplicate dimension name %s", key.Name)
+		}
+		labelNames[key.Name] = struct{}{}
+
+		sanitizedName := sanitize(key.Name)
+		if sanitizedName == key.Name {
+			continue
+		}
+		if _, ok := labelNames[sanitizedName]; ok {
+			return fmt.Errorf("duplicate dimension name %s after sanitization", sanitizedName)
```
Comment (albertteoh) [inline processor/spanmetricsprocessor/processor.go:153]: Please add a test to cover this line.

File: processor/spanmetricsprocessor/processor.go (hunk 1 above)
Comment (yeya24) [reply processor/spanmetricsprocessor/processor.go:422]: Good idea!

File: processor/spanmetricsprocessor/processor.go (hunk 5 above)
Comment (yeya24) [reply processor/spanmetricsprocessor/processor.go:103]: done

File: processor/spanmetricsprocessor/processor.go (hunk 1 above)
Comment (yeya24) [reply processor/spanmetricsprocessor/processor.go:419]: done

File: processor/spanmetricsprocessor/processor.go [hunk 7]
```diff
@@ -379,3 +416,33 @@ func (p *processorImp) cache(serviceName string, span pdata.Span, k metricKey) {
...
+	}
+
+	// Note: No length limit for label keys because Prometheus doesn't
+	// define a length limit, thus we should NOT be truncating label keys.
+	// See https://github.com/orijtech/prometheus-go-metrics-exporter/issues/4.
+	s = strings.Map(sanitizeRune, s)
+	if unicode.IsDigit(rune(s[0])) {
+		s = keyStr + delimiter + s
+	}
+	if s[0] == '_' {
```
[the remark this answers is not in the document]
Comment (yeya24) [reply processor/spanmetricsprocessor/processor.go:431]: Here `_` is a rune while `delimiter` is a string type. For readability, I use string literals now.

File: processor/spanmetricsprocessor/processor.go (hunk 6 above)
Comment (yeya24) [reply processor/spanmetricsprocessor/processor.go:150]: Is this testcase Okay https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2844/files#diff-5a336e1c1dce702d5f77df0fbfc22adde8470d3521537ab9c38d7e2abc7f0d80R539?

**Review by albertteoh — APPROVED** [review]
lgtm @yeya24, thank you!

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**